### ISSUE TITLE
Getting the license file into a Embedded UI in Wpf

### ISSUE THREAD

**Issue by marcobeninca71** [oleg-shilo/wixsharp#313]
Hello everybody
I'm customizing a installation application starting from the example EmbeddedUI_WPF
I'd like to know if it's possible to show in my WPF window the license file which is inside the msi package
thanks in advance

**Comment by oleg-shilo** [comment]
It s actually not as straight forward you would want it to be.
Though WiX guys have created a nice MSI `Session` class to assist with MSI db operations and WixSharp has `GetEmbeddedData` extension. This is what `LicenceDialg` does:
 (https://github.com/oleg-shilo/wixsharp/blob/6c0f2c6fbdf54cd5ec241330e07c6674189066ad/Source/Templates/ProjectTemplates/WixSharp%20Managed%20Setup%20-%20Cusom%20UI/Dialogs/LicenceDialog.cs#L17)
```C#
agreement.Rtf = MsiRuntime.Session.GetEmbeddedString("WixSharp_LicenceFile")
``` 
Basically you add your license file to your setup as a binary asset and at runtime read it via session as byte[] array. Then you convert the array into whatever you want it to be.

**Comment by marcobeninca71** [comment]
Hi Oleg
first of all thanks for your quick reply.
I tried to do like you suggested but it doesn't work. The following lines of code:

string file = MyEmbeddedUI.session.GetEmbeddedString("WixSharp_LicenceFile");
MessageBox.Show(file);

compile because session is a static member of my class MyEmbeddedUI and it's initialized in the function 

public bool Initialize(Session session, string resourcePath, ref InstallUIOptions internalUILevel)

taking it from the first parameter.

but during runtime the MessageBox is never reached, maybe because there's some exception which is not captured by the code.

Is it possible that I missed to include the license file as a binary resource? I use the following code:

project.LicenceFile = @".\\my_license_file.rtf";

thanks

**Comment by marcobeninca71** [comment]
Hi Oleg
I have an update: the issue is how I create the MsiRuntime object. I tried to modify the EmbeddedUI_WPF project adding the followig code:

private MsiRuntime runTime;
 public bool Initialize(Session session, string resourcePath, ref InstallUIOptions internalUILevel)
        {
            runTime = new MsiRuntime(session);
            //....... other stuff as for the example
            return true;
        }

Everything compiles great but trying to run the .msi package it's close immediately. How can I get a valid session object?

I have another question that I think it was already answered but I cannot find it: how can I debug my code in runtime? Can you give me a link to a topic explaining it?

thanks

**Comment by marcobeninca71** [comment]
Sorry for posting too much... anyway I find a very strange behaviour: just adding the following line of code in the variables declarations of the EmbeddedUI_WPF example

public class SampleEmbeddedUI : IEmbeddedUI
    {
        private Thread appThread;
        private Application app;
        private SetupWizard setupWizard;
        private ManualResetEvent installStartEvent;
        private ManualResetEvent installExitEvent;
        **_public ResourcesData UIText = null;_**

makes the msi not working

**Comment by oleg-shilo** [comment]
Timeout please :)

Too many things has been misunderstood. Now, one by one:
 
- `string file = MyEmbeddedUI.session.GetEmbeddedString("WixSharp_LicenceFile");`

   My `GetEmbeddedString` sample was provided not as a solution but as a sample showing the technique of extracting the data from MSI tables.

   You cannot extract "WixSharp_LicenceFile" binary content unless you put it there. And... you didn't.

   I expected you to add your licence file into as a binary MSI asset:
   ```C#
   new Project("MyProduct",
             new Binary(new Id("your_licence "), @".\licence.tf"),
   ```
   And then extract it at runtime from the MSI `Binary` table:
   ```C#
   byte[] data = session.GetEmbeddedData("your_licence");
   ```
- `project.LicenceFile = @".\my_license_file.rtf";`
   This line never embeds the file into MSI but rather creates a dialog definition for the MSI native dialog
with the licence content:
   ![image](https://user-images.githubusercontent.com/16729806/37393189-7f933d48-27c4-11e8-958d-09081562f79f.png)

   You can access `Control` table but it will be easier to use `Binary` table instead.

- `Everything compiles great but trying to run the .msi package it's close immediately. How can I get a valid session object?`

   I am not sure why you need o construct `MsiRuntime`. You have the session object in the `public bool Initialize(Session session,...`. It really has everything that you need.

- how can I debug my code in runtime?
   Put `Assert` in your code and attach the elevated debugger when it asserts. https://github.com/oleg-shilo/wixsharp/wiki/Deployment-scenarios#managed-custom-actions

**Comment by marcobeninca71** [comment]
Dear Olog
thanks very much... now I understand and I was able to run my code!

Just the last question about debugging: I put the Assert and then attach the debugger (Visual Studio 2015) running as administrator. In the case of embedded UI Visual Studio is not able the load the symbols. In the case of your example EmbeddedUI_WPF.pdb is created but the real module in execution id EmbeededUI_WPF.CA.dll and when I try to load manually the symbols they don't match.

Thanks again and compliments for this very useful and complete installation framework

**Comment by oleg-shilo** [comment]
When you embed any NET assembly (e.g. `my_asm.dll`) that is executed at runtime it is embedded (along with it *.pdb) into a special container, which is a **native** dll  (e.g. nt at `my_asm.CA.dl`). Thus your `EmbeededUI_WPF.CA.dll` is not an assembly and it's not expected to have any matching pdb file.

`EmbeededUI_WPF.CA.dll` - in this case is associated with your Custom Action (e.g. WixSharp event or `ManagedCustomAction`). 

At runtime MSI-runtime executes your custom action, which is a special entry point in `EmbeededUI_WPF.CA.dll`. This in turn triggers extraction of the all embedded assemblies, pdb and config files in the temporary folder and eventually the method of the assembly (the one that is marked with `[CustomAction]`) is called. If it compiled with dbg and pdb is found you can do the debugging. If anything this sequence of events did not happen, debugging will not be possible.

I suggest you just try the "DTF_Debugging" sample. It is composed specifically to demonstrate debugging technique.

**Comment by marcobeninca71** [comment]
Dear Oleg
sorry to be late in my answer. The DTF_Debugging example works perfectly but doing the same thing with EmbeddedUI_WPF doesn't do the trick. I don't know why? I tried everything like loading manually the symbols but VS gives me always a mismatch error. Are you able to debug the EmbeddedUI_WPF example? If yes consider closed this thread and I'll try to find a solution on my own

**Comment by oleg-shilo** [comment]
I checked and indeed there is a new negative element in the `msiexec.exe` debugging experience. Not sur is it's not a side effect of a few recent VS updates. 

Though debugging is still possible. THi sis what you need to do differently:

- Place the assert i your UI code
- Build and start your msi
- Observe the assertion prompt. But do not press "Retry".
- In the elevated VS go to attach to process VS dialog.
- Select msiexec.exe that is hosting the assertion dialog. 
- Click "Attach"
- Click "Retry" on the assertion prompt.
- You are in your code now.

![image](https://user-images.githubusercontent.com/16729806/37548280-bf39c6d0-29ca-11e8-87f9-74956a7ea94d.png)

**Comment by marcobeninca71** [comment]
Thanks very much Oleg... debugging works as you suggested

**Comment by oleg-shilo** [comment]
Great. It shouldn't be that convoluted...but it is. 
"Retry->select VS" instance should be enough but it's not. 
I simply discovered it bey "try-n-see" approach.
